### ISSUE TITLE
Compiler panic: Resolving an assert message should happen only once as the VM should have failed

### ISSUE THREAD

**Issue by Pasifaee** [noir-lang/noir#5036]
### Aim

I want to compile the code below. 

Sorry for this code's length and lack of sense. I took code from my codebase and then stripped it down as much as I could and this is a minimal repro I've come up with. Because of this stripping though, it lost its original meaning.

```rust
use dep::std::wrapping_sub;
use dep::std::test::OracleMock;
use dep::std::unsafe::zeroed;

struct Fragment<MAX_DATA_LEN> {
    offset: u64,
    length: u64,
    data: [u8; MAX_DATA_LEN]
}

impl<MAX_DATA_LEN> Fragment<MAX_DATA_LEN> {
    pub fn from_array(data: [u8; MAX_DATA_LEN]) -> Fragment<MAX_DATA_LEN> {
        Fragment { offset: 0, length: data.len() as u64, data }
    }

    pub fn from_vec(vec: BoundedVec<u8, MAX_DATA_LEN>) -> Fragment<MAX_DATA_LEN> {
        Fragment { offset: 0, length: vec.len, data: vec.storage }
    }

    pub fn subfragment(self, length: u64) -> Fragment<MAX_DATA_LEN> {

        Fragment { offset: self.offset, length, data: self.data }
    }

    pub fn at(self, index: u64) -> u8 {
        assert(index < self.length, f"Index {index} out of bounds");
        self.data[self.offset + index]
    }

    pub fn pop_front(&mut self) -> u8 {
        self.offset += 1;
        self.data[self.offset - 1]
    }

    pub fn eq<MAX_DATA_LEN_2>(self, other: Fragment<MAX_DATA_LEN_2>) -> bool {
        let mut res = true;
        for i in 0..MAX_DATA_LEN {
            if (i < self.length) & (i < other.length) {
                let _ = self.at(i);
            }
        }
        res
    }
}

pub fn byte_value<N>(in_value: [u8; N]) -> BoundedVec<u8, N> {
    let mut value_length = 0;

    for i in 0..N {
        let num_bytes_ind = (value_length == 0) as u64;
        let byte_ind = (in_value[i] != 0) as u64;
        value_length = num_bytes_ind * byte_ind;
    }

    BoundedVec { storage: zeroed(), len: value_length }
}

pub fn u64_to_u8(num: u64) -> [u8; 8] {
    let mut out: [u8; 8] = [0; 8];
    for i in 0..8 {
        out[i] = (num >> 0) as u8;
    }

    out
}

struct RlpFragment {
    length: u64,
}

impl RlpFragment {
    fn assert_eq_bytes<FIELD_NAME_LEN, MAX_RLP_LEN, MAX_VALUE_LEN>(
        self,
        rlp: Fragment<MAX_RLP_LEN>,
        value: Fragment<MAX_VALUE_LEN>
    ) {
        let _ = rlp.subfragment(self.length).eq(value);
    }
}

type RlpList<MAX_FIELDS> = BoundedVec<RlpFragment, MAX_FIELDS>;

pub fn decode_to_rlp_header<N>(data: Fragment<N>) -> RlpFragment {
    let mut data = data;
    let prefix = data.pop_front();

    if (prefix < 0x80) {
        RlpFragment { length: 1 }
    } else {
        RlpFragment { length: 0 }
    }
}

pub fn decode_list<N, MAX_FIELDS>(data: Fragment<N>) -> RlpList<MAX_FIELDS> {
    let mut rlp_list: RlpList<MAX_FIELDS> = BoundedVec::new();

    let RlpFragment {length} = decode_to_rlp_header(data);

    let mut current_offset = 0;
    for _ in 0..MAX_FIELDS {
        if (current_offset < length) {
            rlp_list.push(zeroed());
        }
    }

    rlp_list
}

global LEN = 708;

pub fn verify_header(
    number: u64,
    vec: BoundedVec<u8, LEN>
) {
    let header_rlp_list: RlpList<2> = decode_list(Fragment::from_array(vec.storage)); 
    let fragment = Fragment::from_vec(vec);
    header_rlp_list.get(0).assert_eq_bytes(fragment, Fragment::from_vec(byte_value(u64_to_u8(number))));
}

#[oracle(get_header)]
unconstrained fn get_header_oracle() -> (u64, BoundedVec<u8, LEN>) {}

unconstrained fn get_header_unconstrained() -> (u64, BoundedVec<u8, LEN>) {
    get_header_oracle()
}

pub fn get_header() {
    let (number, vec) = get_header_unconstrained();
    verify_header(number, vec);
}

global number = 0;
global data: [u8; LEN] = zeroed();
global bounded_vec = BoundedVec { storage: data, len: 565 };

fn main() {
    let _ = OracleMock::mock("get_header").returns((number, bounded_vec));
    let _ = get_header();
}

#[test]
fn test_main() {
    main();
}
```

### Expected Behavior

The code compiles (and `nargo test` doesn't crash)

### Bug

When running `nargo test` the compiler panics with the following message:

The application panicked (crashed).
Message:  internal error: entered unreachable code: Resolving an assert message should happen only once as the VM should have failed
Location: /Users/runner/work/noir/noir/tooling/nargo/src/ops/execute.rs:148

### To Reproduce

1. Run `nargo test` on the code above 

### Project Impact

Nice-to-have

### Workaround

Yes

### Workaround Description

Replacing `f"Index {index} out of bounds"` in line 26 with `"Index out of bounds"` solves this issue.

### Installation Method

None

### Nargo Version

nargo version = 0.29.0 noirc version = 0.29.0+2da28997dbedce4a00d2162eddb57564bdb23546

### Would you like to submit a PR for this Issue?

None

**Comment by TomAFrench** [comment]
Duplicate of #4804
